### PR TITLE
Functional components and testing-library

### PR DESCRIPTION
**Pull request by ayan4m1**
This PR ports almost all of the existing codebase to use functional components in place of class-based components.

* Add [react-hooks](https://github.com/mixnjuice/frontend/compare/master...ayan4m1:feature/functional-components?expand=1#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5R22) eslint plugin
* Switch to [react-error-boundary](https://github.com/mixnjuice/frontend/compare/master...ayan4m1:feature/functional-components?expand=1#diff-33315431fa6d6994ad75ea830b52e6a5c1aad1bf2c29ecfa32dd3318fb45f7ecR41)
* Remove [ErrorBoundary](https://github.com/mixnjuice/frontend/compare/master...ayan4m1:feature/functional-components?expand=1#diff-43c15e049934eee8256cfff83223109aae08cc70d66763d1a56023545d41527c)
* Add [testing-library/react](https://github.com/mixnjuice/frontend/pull/87/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R68)

### CODE REVIEW

File: src/components/App/App.js [hunk 1]
```diff
@@ -30,67 +29,47 @@ import {
...
-            <Route component={NotFound} />
-          </Switch>
-          <Footer />
-        </ErrorBoundary>
-      </Suspense>
-    );
-  }
+  return (
+    <Suspense fallback={<SuspenseFallback />}>
+      <ErrorBoundary FallbackComponent={ErrorFallback}>
```
Comment (ayan4m1) [inline src/components/App/App.js:41]: This uses the new react-error-boundary component to render the error boundary in a functional way.

File: .eslintrc.js [hunk 2]
```diff
@@ -19,6 +19,7 @@ module.exports = {
     'plugin:jsx-a11y/recommended',
     'plugin:jest/recommended',
     'plugin:import/errors',
+    'plugin:react-hooks/recommended',
```
Comment (ayan4m1) [inline .eslintrc.js:22]: This adds eslint rules for the hooks we're now using.

File: test-setup.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import '@testing-library/jest-dom';
+import '@testing-library/jest-dom/extend-expect';
+import { cleanup } from '@testing-library/react';
+
+afterEach(cleanup);
```
Comment (ayan4m1) [inline test-setup.js:5]: This sets up @testing-library

File: src/sagas/toast.test.js [hunk 4]
```diff
@@ -3,9 +3,7 @@ import { all, put, delay, takeEvery } from 'redux-saga/effects';
 import { actions, types } from 'reducers/toast';
 import toastSaga, { workers, watchers } from 'sagas/toast';
 
-jest.mock('nanoid', () => ({
-  nanoid: jest.fn(() => 'testing')
-}));
+jest.mock('nanoid', () => ({ nanoid: () => 'testing' }));
```
Comment (ayan4m1) [inline src/sagas/toast.test.js:6]: We upgraded to a version of nanoid that doesn't return the same way anymore

File: src/components/IngredientBar/IngredientBar.js [hunk 5]
```diff
@@ -1,109 +1,118 @@
...
+  );
+  const getTooltip = useCallback(
+    (name) => {
+      const { [name]: value } = props;
+      const {
+        [name]: {
+          ref: { current: ref },
+          getter
+        }
+      } = components;
```
Comment (JosefBud) [inline src/components/IngredientBar/IngredientBar.js:44]: I don't know if I'm just tired but this feels like galaxy brain, 4D chess shit

File: src/components/RecipeDetails/RecipeDetails.js [hunk 6]
```diff
@@ -1,76 +1,79 @@
...
+          <span>Max VG |&nbsp;</span>
+        ) : (
+          <span>
+            {percentPg}% PG / {percentVg}% VG |&nbsp;
+          </span>
+        )}
+        {shakeAndVape ? (
+          <span>Shake &amp; Vape |&nbsp;</span>
+        ) : (
+          <span>Steep for {steepDays} days |&nbsp;</span>
```
Comment (JosefBud) [inline src/components/RecipeDetails/RecipeDetails.js:31]: I know this isn't from your PR but just as a general note we should figure out a better way to space this out than using `&nbsp;`

**Review by JosefBud — APPROVED** [review]

File: src/components/App/App.js [hunk 7]
```diff
@@ -30,67 +29,47 @@ import {
...
+      </ErrorBoundary>
+    </Suspense>
+  );
 }
-
-const mapDispatchToProps = (dispatch) => ({
-  actions: bindActionCreators(appActions, dispatch)
-});
-
-export default withRouter(connect(null, mapDispatchToProps)(App));
```
[the remark this answers is not in the document]
Comment (ayan4m1) [reply src/components/App/App.js:96]: I don't think we ever needed it there.

File: src/components/App/App.test.js [hunk 8]
```diff
@@ -1,257 +1,220 @@
...
     }
   });
 
-  afterEach(() => {
-    actions.initApp.mockReset();
-  });
-
   it('calls initApp', () => {
+    const dispatch = useDispatch();
     const RoutedApp = withMemoryRouter(App);
```
[the remark this answers is not in the document]
Comment (ayan4m1) [reply src/components/App/App.test.js:45]: Nah, withRouter only adds props in like location and history if you need them. withMemoryRouter is required for the routing to mount at all.

**Review by daviddyess — APPROVED** [review]
Looks good to me @ayan4m1

**Pull request merged**